### PR TITLE
henry/webrel-319/ts-migration-modules-trading-helpers-folder

### PR DESCRIPTION
**Pull request by henry-deriv**
## Changes:

Please provide a summary of the change.

### Screenshots:

Please provide some screenshots of the change.

### CODE REVIEW

File: packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
+import { PriceProposalResponse } from '@deriv/api-types';
+import { convertToUnix, getDecimalPlaces, getPropertyValue, isAccumulatorContract, toMoment } from '@deriv/shared';
+import { useTraderStore } from 'Stores/useTraderStores';
```
Comment (maryia-deriv) [inline packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts:3]: import { TTradeStore } from 'Types';

File: packages/trader/src/Stores/Modules/Trading/Helpers/process.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import { ContractType as ContractTypeHelper } from 'Stores/Modules/Trading/Helpers/contract-type';
+import * as ContractType from '../Actions/contract-type';
+import * as Duration from '../Actions/duration';
+import * as StartDate from '../Actions/start-date';
+import { useTraderStore } from 'Stores/useTraderStores';
```
Comment (maryia-deriv) [inline packages/trader/src/Stores/Modules/Trading/Helpers/process.ts:5]: https://github.com/kate-deriv/deriv-app/pull/11/files#r1243446315 same )

File: packages/trader/src/Stores/Modules/Trading/Helpers/preview-proposal.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import debounce from 'lodash.debounce';
+import { isEmptyObject, WS } from '@deriv/shared';
+import { createProposalRequests } from './proposal';
+import { useTraderStore } from 'Stores/useTraderStores';
```
Comment (maryia-deriv) [inline packages/trader/src/Stores/Modules/Trading/Helpers/preview-proposal.ts:4]: https://github.com/kate-deriv/deriv-app/pull/11/files#r1243446315 same)

File: packages/trader/src/Types/common-prop.type.ts [hunk 4]
```diff
@@ -14,4 +14,9 @@ export type TProposalTypeInfo = {
     stake: string;
 };
 
+export type TError = {
```
Comment (maryia-deriv) [inline packages/trader/src/Types/common-prop.type.ts:17]: where is this type is used ? )

File: packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    const returns = (profit * 100) / (proposal?.ask_price || 1);
+    const stake = proposal?.display_value;
+    const basis_list = store.basis_list;
+
+    const contract_basis: TObjContractBasis | undefined = store.is_vanilla
+        ? { text: 'Payout', value: 'number_of_contracts' }
+        : basis_list.find(o => o.value !== store.basis);
+
+    const is_stake = contract_basis?.value === 'stake';
+    //@ts-expect-error we are using key that does not exist in proposal
```
Comment (maryia-deriv) [inline packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts:64]: @henry-deriv please use an assertion (as) or a type guard (if conditions) in places, where you have TS complaining about the types it's not aware about, instead of this comment.

File: packages/trader/src/Stores/Modules/Trading/Helpers/limit-orders.ts [hunk 6]
```diff
@@ -51,7 +67,8 @@ export const setLimitOrderBarriers = ({ barriers, contract_type, contract_info =
                     isSingleBarrier: true,
                     opacityOnOverlap: key === LIMIT_ORDER_TYPES.STOP_OUT && 0.15,
                 };
-                barrier = new ChartBarrierStore(obj_limit_order.value);
+                //@ts-expect-error until chart barrier store in modules/smartchart is migrated to TS
```
Comment (maryia-deriv) [inline packages/trader/src/Stores/Modules/Trading/Helpers/limit-orders.ts:70]: https://github.com/kate-deriv/deriv-app/pull/11/files#r1243453750 same)

**Review by maryia-deriv — CHANGES_REQUESTED** [review]
great job! please check out some suggestions. please also declare variables for ReturnType<typeof useStore> and ReturnType<typeof useTraderStore> when you use them multiple times in one file. 🙏🏻

File: packages/trader/src/Types/common-prop.type.ts (hunk 4 above)
Comment (henry-deriv) [reply packages/trader/src/Types/common-prop.type.ts:17]: i saw it and copied it from ur one of your PRs

File: packages/trader/src/Stores/Modules/Trading/Helpers/preview-proposal.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import { PriceProposalResponse } from '@deriv/api-types';
+
+export const requestPreviewProposal = debounce(
+    (
+        store: ReturnType<typeof useTraderStore>,
+        override = {},
+        onProposalResponse: (response: PriceProposalResponse) => void
+    ) => {
+        const new_store = { ...store, ...override };
+        const requests = createProposalRequests(new_store) as Record<string, unknown>;
```
[the remark this answers is not in the document]
Comment (henry-deriv) [reply packages/trader/src/Stores/Modules/Trading/Helpers/preview-proposal.ts:14]: worked!

File: packages/trader/src/Stores/Modules/Trading/Helpers/process.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+const processInSequence = async (
+    store: ReturnType<typeof useTraderStore>,
+    functions: ReturnType<typeof getMethodsList>
+) => {
+    //@ts-expect-error error will persist until basestore is typescript converted
```
[the remark this answers is not in the document]
Comment (henry-deriv) [reply packages/trader/src/Stores/Modules/Trading/Helpers/process.ts:11]: resolved others but i dont know how to get around this one

[the remark this answers is not in the document]
Comment (henry-deriv) [reply packages/trader/src/Stores/Modules/Trading/Helpers/process.ts]: So it turns out the error that is left cannot be gotten rid of until basestore is typescript migration and make input parameters optional. It's expecting `string[]` and putting `['']` breaks the whole application. Better to leave it as ts-expect error.

File: packages/stores/src/mockStore.ts [hunk 9]
```diff
@@ -316,7 +316,18 @@ const mock = (): TStores & { is_mock: boolean } => {
         portfolio: {
             active_positions: [],
             all_positions: [],
-            error: '',
+            error: {
```
Comment (kate-deriv) [inline packages/stores/src/mockStore.ts:319]: Do we need this object here? Isn't it just a string? I guess error is object in common store, not in portfolio store

Comment (henry-deriv) [reply packages/stores/src/mockStore.ts:319]: You are probably right. These changes came in when i pulled ur branch though. I didn't add this error object in mock

Comment (henry-deriv) [reply packages/stores/src/mockStore.ts:319]: Actually no. i think it needs to be defined where ever TCommonStoreError exists

Comment (kate-deriv) [reply packages/stores/src/mockStore.ts:319]: But error in portfolio store is string and TCommonStoreError is for error from common store, isn't it?

Comment (henry-deriv) [reply packages/stores/src/mockStore.ts:319]: no. They are both TCommonStoreError now. At least thats what i got from pulling ur branch

Comment (maryia-deriv) [reply packages/stores/src/mockStore.ts:319]: @henry-deriv error in portfolio-store is always a string. not the same as in common-store. please revert your changes here. 🙏🏻 
Screenshot 2023-07-05 at 10 05 41 AM
Screenshot 2023-07-05 at 10 06 32 AM
Screenshot 2023-07-05 at 10 07 01 AM

File: packages/stores/types.ts [hunk 10]
```diff
@@ -335,7 +335,7 @@ type TPortfolioPosition = {
 type TPortfolioStore = {
     active_positions: TPortfolioPosition[];
     all_positions: TPortfolioPosition[];
-    error: string;
+    error: TCommonStoreError;
```
Comment (maryia-deriv) [inline packages/stores/types.ts:338]: @henry-deriv please revert this change, explanation is here https://github.com/kate-deriv/deriv-app/pull/11/files#r1252641368

File: packages/trader/src/Stores/Modules/Trading/Helpers/process.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+import * as ContractType from '../Actions/contract-type';
+import * as Duration from '../Actions/duration';
+import * as StartDate from '../Actions/start-date';
+import { TTradeStore } from 'Types';
+
+const processInSequence = async (
+    store: TTradeStore,
+    functions: ReturnType<typeof getMethodsList> | ReturnType<typeof getExpiryMethodsList>
+) => {
+    //@ts-expect-error need to typescript convert base store and make parameters optional else it will break
```
Comment (maryia-deriv) [inline packages/trader/src/Stores/Modules/Trading/Helpers/process.ts:11]: @henry-deriv let's use assertions (as ) or a type guard at the place of error instead of this `//@ts-expect-error` comment.

File: packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts [hunk 12]
```diff
@@ -18,60 +45,69 @@ export const getProposalErrorField = response => {
...
-    const proposal = response.proposal || {};
-    const profit = proposal.payout - proposal.ask_price || 0;
-    const returns = (profit * 100) / (proposal.ask_price || 1);
-    const stake = proposal.display_value;
+export const getProposalInfo = (
+    store: TTradeStore,
+    response: PriceProposalResponse & TError,
+    obj_prev_contract_basis: TObjContractBasis
+) => {
+    const proposal = response.proposal || ({} as PriceProposalResponse['proposal']);
```
Comment (maryia-deriv) [inline packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts:53]: PriceProposalResponse['proposal'] is used multiple times in this file, so it can be replaced with a shorter `Proposal` type which can be imported from api-types as well :)

File: packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts [hunk 13]
```diff
@@ -18,60 +45,69 @@ export const getProposalErrorField = response => {
...
+    const profit = (proposal?.payout || 0) - (proposal?.ask_price || 0);
+    const returns = (profit * 100) / (proposal?.ask_price || 1);
+    const stake = proposal?.display_value;
     const basis_list = store.basis_list;
 
-    const contract_basis = store.is_vanilla
+    const contract_basis: TObjContractBasis | undefined = store.is_vanilla
         ? { text: 'Payout', value: 'number_of_contracts' }
-        : basis_list.find(o => o.value !== store.basis) || {};
+        : basis_list.find(o => o.value !== store.basis);
```
Comment (maryia-deriv) [inline packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts:61]: let's leave `|| {}` here in case the item is not found.

File: packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts [hunk 14]
```diff
@@ -18,60 +45,69 @@ export const getProposalErrorField = response => {
...
 
     const obj_contract_basis = {
-        text: contract_basis.text || '',
+        text: contract_basis?.text || '',
         value: price || '',
     };
 
-    const commission = proposal.commission;
-    const cancellation = proposal.cancellation;
+    const commission = proposal?.commission;
```
Comment (maryia-deriv) [inline packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts:81]: @henry-deriv `?` marks after `proposal` should not be necessary because it cannot be undefined. it will be assigned with `{}` at line 53 if there is no `proposal` in response.

Comment (henry-deriv) [reply packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts:81]: removing `?mark` throws ts error. I added it due to that. I just tried again. Same error persists.

File: packages/trader/src/Stores/Modules/Trading/Helpers/process.ts (hunk 11 above)
Comment (henry-deriv) [reply packages/trader/src/Stores/Modules/Trading/Helpers/process.ts:11]: i already tried @maryia-deriv

the issue is `getSnapshot` expects an input of `string[]`. So i tried getSnapshot(['']) to get rid of the error but breaks the application. So i cant do anything until getSnapshot is ts migrated to make its parameters optional.

File: packages/stores/types.ts (hunk 10 above)
Comment (henry-deriv) [reply packages/stores/types.ts:338]: reverted

File: packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts [hunk 15]
```diff
@@ -18,60 +45,69 @@ export const getProposalErrorField = response => {
...
+export const getProposalInfo = (
+    store: TTradeStore,
+    response: PriceProposalResponse & TError,
+    obj_prev_contract_basis: TObjContractBasis
+) => {
+    const proposal = response.proposal || ({} as Proposal);
+    const profit = (proposal?.payout || 0) - (proposal?.ask_price || 0);
+    const returns = (profit * 100) / (proposal?.ask_price || 1);
+    const stake = proposal?.display_value;
     const basis_list = store.basis_list;
```
Comment (maryia-deriv) [inline packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts:57]:
```suggestion
    const basis_list: Record<'text' | 'value', string>[] = store.basis_list;
```

File: packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts (hunk 14 above)
Comment (maryia-deriv) [reply packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts:81]: @henry-deriv which error are you getting?  I've removed all `?` marks and haven't got any error 🙂 :  
Screenshot 2023-07-05 at 11 14 34 AM

Comment (henry-deriv) [reply packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts:81]: oh weird. It was just there. Not anymore though!

File: packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts (hunk 15 above)
Comment (henry-deriv) [reply packages/trader/src/Stores/Modules/Trading/Helpers/proposal.ts:57]: Screenshot 2023-07-05 at 4 19 58 PM

Already implied 👍

**Review by maryia-deriv — APPROVED** [review]

**Review by kate-deriv — APPROVED** [review]
